FaunaDB.Client/Encoding/Encoder.cs

```diff
@@ -0,0 +1,258 @@
+
+            var properties = srcType
+                .GetProperties()
+                .Where(parameter => parameter.CanRead && !Has<IgnoreAttribute>(parameter))
+                .Select(parameter => AddElementToDic(encoderExpr, srcType, parameter, addMethod, objExpr));
+
+            var newDictExpr = Expression.ListInit(Expression.New(dictType), fields.Concat(properties));
+            var lambda = Expression.Lambda<Converter>(
+                Expression.New(objectVConstructor, newDictExpr),
+                encoderExpr,
+                objExpr
+            );
+
+            return lambda.Compile();
+        }
+
+        /*
+         * (object) ((SrcType)objExpr).member
+         */
+        Expression AccessMember(Type srcType, MemberInfo member, Expression objExpr)
+        {
+            return Expression.Convert(
+                Expression.MakeMemberAccess(Expression.Convert(objExpr, srcType), member),
+                typeof(object)
+            );
+        }
+
+        /*
+         * encodeExpr.Encode( argExpression )
+         */
+        Expression CallEncode(Expression encoderExpr, Expression argExpression)
+        {
+            var encodeMethod = typeof(EncoderImpl).GetMethod("Encode", new Type[] { typeof(object) });
+
+            return Expression.Call(
+                encoderExpr,
+                encodeMethod,
+                argExpression
+            );
+        }
+
+        /*
+         * Add(member.Name, Encode( objExpr.member ))
+         */
+        ElementInit AddElementToDic(Expression encoderExpr, Type srcType, MemberInfo member, MethodInfo addMethod, Expression objExpr)
+        {
+            var keyExpr = Expression.Constant(GetMemberName(member));
+            var valueExpr = CallEncode(encoderExpr, AccessMember(srcType, member, objExpr));
+
+            return Expression.ElementInit(
+                addMethod, keyExpr, valueExpr
+            );
+        }
+
+        static bool Has<T>(MemberInfo member) where T : Attribute =>
+            member.GetCustomAttribute<T>() != null;
+
+        static IEnumerable<FieldInfo> GetAllFields(Type type)
+        {
+            if (type == null)
+                return Enumerable.Empty<FieldInfo>();
+
+            var flags = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;
+
+            return type.GetFields(flags).Concat(GetAllFields(type.BaseType));
+        }
+
+        static string GetMemberName(MemberInfo member)
+        {
+            var field = member.GetCustomAttribute<FieldAttribute>();
+
+            if (field != null)
+                return field.Name;
+
+            return member.Name;
+        }
+
+        class ReferenceComparer : IEqualityComparer<object>
+        {
+            public static readonly ReferenceComparer Default = new ReferenceComparer();
+
+            public new bool Equals(object x, object y) =>
+                object.ReferenceEquals(x, y);
```

Comment:
That shouldn't be a problem specially for strings, because it's a primitive type so we don't recurse into the struct/class fields. Also even if we use the very same reference in two or more fields of a class again it won't be a problem because it will encode each field recursively and when it return the recursion it removes the reference from the stack.